File: packages/sdk-py/src/grantex/resources/_compliance.py
from __future__ import annotations

from urllib.parse import urlencode

from .._http import HttpClient
from .._types import (
    ComplianceAuditExport,
    ComplianceExportAuditParams,
    ComplianceExportGrantsParams,
    ComplianceGrantsExport,
    ComplianceSummary,
)


class ComplianceClient:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def get_summary(
        self,
        *,
        since: str | None = None,
        until: str | None = None,
    ) -> ComplianceSummary:
        """Get an org-wide compliance summary (agents, grants, audit, policies, plan)."""
        params: dict[str, str] = {}
        if since is not None:
            params["since"] = since
        if until is not None:
            params["until"] = until
        qs = urlencode(params)
        path = f"/v1/compliance/summary?{qs}" if qs else "/v1/compliance/summary"
        data = self._http.get(path)
        return ComplianceSummary.from_dict(data)

    def export_grants(
        self,
        params: ComplianceExportGrantsParams | None = None,
    ) -> ComplianceGrantsExport:
        """Export all grants (optionally filtered)."""
        qs = urlencode(params.to_dict()) if params else ""
        path = f"/v1/compliance/export/grants?{qs}" if qs else "/v1/compliance/export/grants"
        data = self._http.get(path)
        return ComplianceGrantsExport.from_dict(data)

    def export_audit(
        self,
        params: ComplianceExportAuditParams | None = None,
    ) -> ComplianceAuditExport:
        """Export all audit entries (optionally filtered)."""
        qs = urlencode(params.to_dict()) if params else ""
        path = f"/v1/compliance/export/audit?{qs}" if qs else "/v1/compliance/export/audit"
        data = self._http.get(path)
        return ComplianceAuditExport.from_dict(data)

File: packages/sdk-ts/src/resources/compliance.ts
import type { HttpClient } from '../http.js';
import type {
  ComplianceAuditExport,
  ComplianceExportAuditParams,
  ComplianceExportGrantsParams,
  ComplianceGrantsExport,
  ComplianceSummary,
  EvidencePack,
  EvidencePackParams,
} from '../types.js';

export class ComplianceClient {
  readonly #http: HttpClient;

  constructor(http: HttpClient) {
    this.#http = http;
  }

  /** Get an org-wide compliance summary (agents, grants, audit, policies, plan). */
  getSummary(params?: { since?: string; until?: string }): Promise<ComplianceSummary> {
    const query = buildQuery(params);
    const path = query ? `/v1/compliance/summary?${query}` : '/v1/compliance/summary';
    return this.#http.get<ComplianceSummary>(path);
  }

  /** Export all grants (optionally filtered). */
  exportGrants(params?: ComplianceExportGrantsParams): Promise<ComplianceGrantsExport> {
    const query = buildQuery(params);
    const path = query
      ? `/v1/compliance/export/grants?${query}`
      : '/v1/compliance/export/grants';
    return this.#http.get<ComplianceGrantsExport>(path);
  }

  /** Export all audit entries (optionally filtered). */
  exportAudit(params?: ComplianceExportAuditParams): Promise<ComplianceAuditExport> {
    const query = buildQuery(params);
    const path = query
      ? `/v1/compliance/export/audit?${query}`
      : '/v1/compliance/export/audit';
    return this.#http.get<ComplianceAuditExport>(path);
  }

  /** Generate a full SOC2/GDPR evidence pack with chain integrity verification. */
  evidencePack(params?: EvidencePackParams): Promise<EvidencePack> {
    const query = buildQuery(params);
    const path = query
      ? `/v1/compliance/evidence-pack?${query}`
      : '/v1/compliance/evidence-pack';
    return this.#http.get<EvidencePack>(path);
  }
}

function buildQuery(params?: object): string {
  if (!params) return '';
  const entries = (Object.entries(params) as Array<[string, unknown]>).filter(
    ([, v]) => v !== undefined && v !== null,
  ) as Array<[string, string]>;
  if (entries.length === 0) return '';
  return entries
    .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(String(v))}`)
    .join('&');
}
